Add connector development process (#6353)
ballerina-platform/ballerina-standard-library: docs/connector-development-process.md

# Connector Development Process

# Library Development Process 
_Author_: @shafreenAnfar @ThisaruGuruge @NipunaRanasinghe @niveathika @TharmiganK    
_Reviewer_: @daneshk @ThisaruGuruge @NipunaRanasinghe  
_Created_: 2024/04/17   
_Created_: 2024/04/17   

## Overview

This section discusses the connector development process from inception to realization. The goal is to ensure we deliver the best possible connector experience. The sequence of each step might be adjusted slightly, but the overall process remains unchanged. Once a library completes each step and reaches the end, we can designate it as a generally available connector. Each connector owner is responsible for ensuring their libraries undergo these steps before signing off.

## Naming Convention

This section introduces the naming conventions you need to be aware of before starting the development of connectors.

> Package:

Reviewer: ```suggestion
### Package
```

- ballerinax/salesforce.commons

### Hierarchical Packages vs. Submodules

#### Hierarchical Packages
Hierarchical package names are preferred when there is a need to distinguish between connectors related to the same vendor/provider. For example, the separation between `azure.eventhub` and `azure.servicebus` can be achieved through hierarchical packages, making it clear that these two are related to the same provider, yet they represent distinct services. It's also recommended to have multiple packages with hierarchical naming when dealing with multiple APIs related to the same SaaS product having independent versioning and release cycles. For example, Salesforce supports multiple APIs having independent API versioning. Therefore, it is recommended to have multiple connectors, each in its separate hierarchical package:

- ballerinax/salesforce.bulk
- ballerinax/salesforce.soap

#### Submodules
Submodules should primarily be used to organize the code structure within a single connector package. Avoid defining multiple sub-clients in each submodule within a single connector package, as this may lead to code ambiguity and complexity.

### Vendor/Platform Prefixes

> Generic SaaS Product Names:

Reviewer: ```suggestion
#### Generic SaaS Product Names
```

When dealing with SaaS product names that are more generic and not specific to a particular vendor, it is recommended to add a vendor/platform abbreviation as a prefix to the package name. This helps to distinguish the client usages and avoid ambiguity, especially when multiple vendors offer similar services. For example, Azure provides storage services. In this case, using a vendor/platform abbreviation as a prefix is advisable:

- ballerinax/azure.azstorage

> ### Vendor-Specific Names

Reviewer: ```suggestion
#### Vendor-Specific Names
```

If the SaaS product name is specific to a particular vendor and is unlikely to cause conflicts or confusion with other connector usages in the same program, there may be no need to prefix the package name with a vendor/platform abbreviation. For example, Azure provides CosmosDB. In this case, using the connector as ‘cosmosdb:Client’ is less likely to create conflicts or confusion with other connectors:

- ballerinax/azure.cosmosdb

> ### Summary

Reviewer: ```suggestion
#### Summary
```

Package names should be clear and descriptive, making it easy for developers to understand the purpose of each package and module. Good package names are typically short, clear, and comprised of simple nouns.

Despite being syntactically correct, some naming patterns that are typical in other languages might not be idiomatic in Ballerina. For example, consider the following package names, which may be recommended in other languages but do not fit well in Ballerina:

- azureEventHub
- azure_event_hub

It is acceptable to use abbreviations as the package name if the abbreviation is more familiar among the developer community. Most widely-used packages often have compressed names. However, using any cryptic abbreviations or overly generic names should be avoided.

#### Consistency

Maintain consistency in naming throughout the connector. Consistency promotes a unified and intuitive experience for developers using the connector.

#### Avoiding Overly Deep Hierarchies

While nesting subpackages can be useful for organization, avoid creating overly deep hierarchies. Packages should provide a logical structure without becoming excessively complex.

#### Avoiding Split-Module Conditions

> A “split-module condition” can occur when the latest versions of two different packages contain the same module, resulting in a build failure. Therefore, hierarchical structures should be chosen carefully and used with caution to minimize such situations. (Refer to Ballerina's documentation on split-module conditions for more details).

Reviewer: ```suggestion
A “split-module condition” can occur when the latest versions of two different packages contain the same module, resulting in a build failure. Therefore, hierarchical structures should be chosen carefully and used with caution to minimize such situations. (Refer to [Ballerina package references documentation](https://ballerina.io/learn/package-references/#the-split-module-condition) for more details).
```

### Write Specification

You must write a specification for your connector. In the case of REST APIs, your specification can be the OAS (OpenAPI Specification). However, if you have made any alterations, or as we call them, sanitations, you must document them in sanitations.md. Additionally, you may include a `spec.md`.

### Test Consideration

The code we write for connectors differs significantly from the code we write for other modules. Most connectors are generated using the OpenAPI tool, and some are handwritten. Handwritten connectors simply wrap a given SDK to create the Ballerina version of it. The code required for this is very minimal, lacks mutable state, and each operation is independent of the others.

With the above in mind, the following are the aspects covered by test cases:

- Validate the implementation against the contract (not common),
- Catch any regression issues (not common),
- Catch any breaking changes introduced by language changes (relatively common),
- Validate against platforms such as GraalVM, Java11, etc. (not common).

> Need for Tests

Reviewer 2: ```suggestion

#### Need for Tests
```
